wolfCLU/src/crypto/clu_crypto_setup.c

```diff
@@ -0,0 +1,358 @@
+    int      outCheck   =   0;  /* if output has been provided */
+    int      i          =   0;  /* loop counter */
+    int      eCheck     =   0;  /* if user is encrypting data */
+    int      dCheck     =   0;  /* if user is decrypting data */
+    int      inputHex   =   0;  /* if user is encrypting hexidecimal stuff */
+    int      keyType    =   0;  /* tells Decrypt which key it will be using
+                                 * 1 = password based key, 2 = user set key
+                                 */
+    int      printOut   =   0;  /* flag to print out key/iv/salt */
+    int      pbkVersion =   1;
+    const WOLFSSL_EVP_MD* hashType = wolfSSL_EVP_sha256();
+
+    const WOLFSSL_EVP_CIPHER* cphr = NULL;
+    word32   ivSize     =   0;  /* IV if provided should be 2*block */
+    word32   numBits    =   0;  /* number of bits in argument from the user */
+    int      option;
+    int      long_index = 0;
+
+    if (action == 'e')
+        eCheck = 1;
+    if (action == 'd')
+        dCheck = 1;
+
+    ret = wolfCLU_checkForArg("-h", 2, argc, argv);
+    if (ret > 0) {
+        if (eCheck == 1) {
+            wolfCLU_encryptHelp();
+            return 0;
+        } else {
+            wolfCLU_decryptHelp();
+            return 0;
+        }
+    }
+
+    /* gets blocksize, algorithm, mode, and key size from name argument */
+    block = wolfCLU_getAlgo(argc, argv, &alg, &mode, &size);
+    if (block < 0) {
+        printf("unable to find algorithm to use\n");
+        return -1;
+    }
+
+    /* initialize memory buffers */
+    pwdKey = (byte*) XMALLOC(size + block, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
+    if (pwdKey == NULL)
+        return MEMORY_E;
+    XMEMSET(pwdKey, 0, size + block);
+
+    iv = (byte*) XMALLOC(block, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
+    if (iv == NULL) {
+        wolfCLU_freeBins(pwdKey, NULL, NULL, NULL, NULL);
+        return MEMORY_E;
+    }
+    XMEMSET(iv, 0, block);
+
+    key = (byte*) XMALLOC(size, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
+    if (key == NULL) {
+        wolfCLU_freeBins(pwdKey, iv, NULL, NULL, NULL);
+        return MEMORY_E;
+    }
+    XMEMSET(key, 0, size);
+
+    opterr = 0; /* do not display unrecognized options */
+    optind = 0; /* start at indent 0 */
+    while ((option = getopt_long_only(argc, argv, "",
+                   crypt_options, &long_index )) != -1) {
+
+        switch (option) {
+        case WOLFCLU_PASSWORD: /* Password                                  */
+            XSTRNCPY((char*)pwdKey, optarg, size);
+            pwdKeyChk = 1;
+            keyType   = 1;
+            break;
+
+        case WOLFCLU_PBKDF2:
+            pbkVersion = 2;
+            break;
+
+        case WOLFCLU_KEY:      /* Key if used must be in hex                */
+            break;
+
+        case WOLFCLU_IV:       /* IV if used must be in hex                 */
+            {
+                char ivString[XSTRLEN(optarg)];
```

Comment:
Once again I don't get how this works. Doesn't this have to be dynamically allocated?